src/Broker/Broker.php

```diff
@@ -470,11 +478,26 @@ public function hasConstant(\PhpParser\Node\Name $nameNode, ?Scope $scope): bool
 
 	public function resolveConstantName(\PhpParser\Node\Name $nameNode, ?Scope $scope): ?string
 	{
-		return $this->resolveName($nameNode, function (string $name): bool {
+		return $this->resolveName($nameNode, function (string $name) use ($scope): bool {
+			$isCompilerHaltOffset = $name === '__COMPILER_HALT_OFFSET__';
+			$compilerHaltOffsetIsCalled = $scope !== null && $this->fileHasCompilerHaltStatementCalls($scope->getFile());
+			if ($isCompilerHaltOffset && $compilerHaltOffsetIsCalled) {
+				return true;
+			}
 			return defined($name);
 		}, $scope);
 	}
 
+	private function fileHasCompilerHaltStatementCalls(string $pathToFile): bool
+	{
+		$nodes = $this->parser->parseFile($pathToFile);
+		$haltCompilerCalls = array_filter($nodes, function (Node $node) {
```

Comment:
1) This could be more efficient - stop after first occurence, don't go through all nodes.
2) How exactly the halt compiler call works? Can it be nested inside an if or some other construct or is it always in the top level?

Reply:
> This could be more efficient - stop after first occurence, don't go through all nodes.
Indeed.

> How exactly the halt compiler call works? Can it be nested inside an if or some other construct or is it always in the top level?

No it cannot be nested. As everything after `__halt_compiler()` is ignored by compiler. So if you try use it inside an `if` with opening curly bracket it will lead to invalid syntax cause closing bracket will be missing for compiler.

Few examples: 
https://3v4l.org/rWSHt
https://3v4l.org/oi3F4

So, is there a way to analyze only top-level nodes?

Reply:
Yes, this is how you did it 😊

Reply:
Hah, I didn't realized it 😄